modernize nix a little

michaelglass opened the pull request:
# the problems
1. updating (or changing) the shell's veresion of `librdkafka` from nix is an arduous process
2. `nix-shell` is slow to launch

# the solutions
## 1. [`niv`](https://github.com/nmattia/niv) handles nix versioning.

### before
1. visit `https://github.com/nixos/nixpkgs`
2. find the right branch
3. find the right rev
4. download the rev and get the SHA of the download
5. update both in the right nix file.

### after
`$ niv update`

## 2. [opt-in] if you have direnv installed and setup, `direnv allow` will cache nix-shell launch
this PR adds a `.envrc`, which in combination with `direnv` makes nix usage snappy

it also updates change the existing `nixpkgs` version to resolve security warnings with modern nixpkgs.

AlexeyRaga commented:
Thanks, I'll have a look soon.
The diff looks scary though ;) We only need to provide _a_ version of `librdkafka` in shell, and having that much nix code for it... Not sure how I feel about it, but I'll have a look :)

michaelglass commented:
one of the files is the generated json parser from niv

michaelglass commented:
@AlexeyRaga
- updated the PR description to explain the intent
- split up the commits so it's clear what part of this code was auto-generated by `niv`

endgame reviewed (approved):
@AlexeyRaga All this is doing is pinning nixpkgs instead of getting whatever version is in the user's channels. Until flakes take over (eventually, I hope, because they seem good in theory), `niv` seems to be the predominant way of pinning remote nix deps.

AlexeyRaga commented:
@michaelglass There was a bit of a mess with branches (GH force-switched the default branch from `master` to `main` and I haven't noticed). I have update `main`, could you please update your PR too?

michaelglass commented:
thanks @AlexeyRaga. rebased on main
